File: creational/dart/abstract_factory_structured/concrete_products.dart
import 'interfaces.dart';

class WindowsButton extends Button {
  @override
  void click() => print('Windows button clicked!');
  @override
  void render() => print('Rendering Windows-style button...');
  @override
  String get name => 'Windows Button';
}

class WindowsCheckbox extends Checkbox {
  @override
  void check() => print('Windows checkbox checked!');
  @override
  void render() => print('Rendering Windows-style checkbox...');
  @override
  String get name => 'Windows Checkbox';
}

class MacButton extends Button {
  @override
  void click() => print('Mac button clicked!');
  @override
  void render() => print('Rendering Mac-style button...');
  @override
  String get name => 'Mac Button';
}

class MacCheckbox extends Checkbox {
  @override
  void check() => print('Mac checkbox checked!');
  @override
  void render() => print('Rendering Mac-style checkbox...');
  @override
  String get name => 'Mac Checkbox';
}

class MySQLConnection extends IDatabaseConnection {
  @override
  void connect() => print('Connecting to MySQL database...');
  @override
  void disconnect() => print('Disconnecting from MySQL database...');
  @override
  void executeQuery(String query) => print('Executing MySQL query: $query');
}

class MySQLCommand extends DatabaseCommand {
  final String _command;
  MySQLCommand(this._command);
  @override
  void execute() => print('Executing MySQL command: $_command');
}

class PostgreSQLConnection extends IDatabaseConnection {
  @override
  void connect() => print('Connecting to PostgreSQL database...');
  @override
  void disconnect() => print('Disconnecting from PostgreSQL database...');
  @override
  void executeQuery(String query) => print('Executing PostgreSQL query: $query');
}

class PostgreSQLCommand extends DatabaseCommand {
  final String _command;
  PostgreSQLCommand(this._command);
  @override
  void execute() => print('Executing PostgreSQL command: $_command');
}

File: creational/typescript/abstract_factory_structured/concrete_products.ts
import { Button, Checkbox, IDatabaseConnection, DatabaseCommand } from './interfaces';

export class WindowsButton implements Button {
  click(): void { console.log('Windows button clicked!'); }
  render(): void { console.log('Rendering Windows-style button...'); }
  getName(): string { return 'Windows Button'; }
}

export class WindowsCheckbox implements Checkbox {
  check(): void { console.log('Windows checkbox checked!'); }
  render(): void { console.log('Rendering Windows-style checkbox...'); }
  getName(): string { return 'Windows Checkbox'; }
}

export class MacButton implements Button {
  click(): void { console.log('Mac button clicked!'); }
  render(): void { console.log('Rendering Mac-style button...'); }
  getName(): string { return 'Mac Button'; }
}

export class MacCheckbox implements Checkbox {
  check(): void { console.log('Mac checkbox checked!'); }
  render(): void { console.log('Rendering Mac-style checkbox...'); }
  getName(): string { return 'Mac Checkbox'; }
}

export class MySQLConnection implements IDatabaseConnection {
  connect(): void { console.log('Connecting to MySQL database...'); }
  disconnect(): void { console.log('Disconnecting from MySQL database...'); }
  executeQuery(query: string): void { console.log(`Executing MySQL query: ${query}`); }
}

export class MySQLCommand implements DatabaseCommand {
  constructor(private command: string) {}
  execute(): void { console.log(`Executing MySQL command: ${this.command}`); }
}

export class PostgreSQLConnection implements IDatabaseConnection {
  connect(): void { console.log('Connecting to PostgreSQL database...'); }
  disconnect(): void { console.log('Disconnecting from PostgreSQL database...'); }
  executeQuery(query: string): void { console.log(`Executing PostgreSQL query: ${query}`); }
}

export class PostgreSQLCommand implements DatabaseCommand {
  constructor(private command: string) {}
  execute(): void { console.log(`Executing PostgreSQL command: ${this.command}`); }
}
